File: Year-2023/LeetCode-February-Challenge/FindAllAnagramsinaString.js
var findAnagrams = function(s, p) {
    let lenS = s.length;
    let lenP = p.length;
    if(lenP > lenS) return [];

    let count = new Array(26).fill(0);

    for(let i = 0; i < lenP; ++i){
        count[p.charCodeAt(i) - 97]++;
        count[s.charCodeAt(i) - 97]--;
    }
    let res = [];
    if(!count.some(e => e !== 0)) res.push(0);

    for(let i = lenP; i < lenS; ++i){
        count[s.charCodeAt(i - lenP) - 97]++;
        count[s.charCodeAt(i) - 97]--;
        if(!count.some(e => e !== 0)) res.push(i - lenP + 1);
    }
    return res;
};
let  s = "cbaebabacd", p = "abc";
console.log(findAnagrams(s, p))

File: Year-2023/LeetCode-February-Challenge/FindAllAnagramsinaString.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int lenS = s.size();
        int lenP = p.size();
        if(lenP > lenS) return {};

        vector<int> pFreq(26, 0);
        vector<int> window(26, 0);

        for(int i = 0; i < lenP; ++i){
            pFreq[p[i] - 'a']++;
            window[s[i] - 'a']++;
        }
        vector<int> res;
        if(pFreq == window) res.push_back(0);

        for(int i = lenP; i < lenS; ++i){
            window[s[i - lenP] - 'a']--;
            window[s[i] - 'a']++;
            if(pFreq == window) res.push_back(i - lenP + 1);
        }
        return res;
    }
};
int main(){
    string s = "cbaebabacd", p = "abc";
    Solution *S = new Solution();
    vector<int> res = S->findAnagrams(s, p);
    for(auto ch: res) cout<<ch<<" ";
    cout<<endl;
    return 0;
}